### PR TITLE
Add channel invitations

### LINKED ISSUE #1400: Allow moderators to add new members to private channels
As a moderator, I'd like to be able to add new users by email to my private channel. But currently, I can't add anyone who doesn't have an account. 

We can do this by allowing unrecognized emails in the channel members form. When the email address doesn't match an existing user, we can generate an email with a custom message, but that functions similarly to 

### acceptance criteria
- [x] draft text for invite email: https://zpl.io/VQ9RBe9
- [ ] update the `/signup/details` template text so that it will work with this flow
- [ ] determine how best to temporarily store the account information

### Questions
- Can we just create an account, without it being active?
  **A:** No. This won't work the social auth workflow. We would have an account with no social auths, so it would be impossible to log in to the account.
- Can we allow the new user to sign up with any email address or do they have to sign up with the invited email address?
  **A:** They have to sign up with the invited email address
- What do we do if multiple invites have been sent to the same user (this might affect the answer to the Q directly above this)?
  **A:** I think the previous answer makes this moot?
- Do we want to normalize the channel invitation experience across users that do and don't exist?
  - i.e. are we going to allow any channel moderator to add any user, hence permitting them to get content on the frontpage of arbitrary users? (this is probably not a huge issue right now, but is more critical if we open up who can create channels)
  **A:** No plans now. We'll keep an eye on this. 
- Do we need to add the ability to unsubscribe from channels too?
  **A:** We will be adding this soon.

### PR DESCRIPTION
**Pull request by rhysyngsun**
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
  - [x] Mobile width screenshots
  - [x] Tag @ferdi or @pdpinch for review
- [x] Migrations
  - [x] Migration is backwards-compatible with current production code
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Fixes #1400 

#### What's this PR do?

Adds channel invitation functionality

#### How should this be manually tested?
- Go to the contributors tab of a private channel and enter an email address that does not exist as a user.
- You should receive an email that prompts you to register.
- Registering with that email should resolve outstanding invitations into channel memberships

Known Limitations:
- Registering with a different email will not carry over the channel invitation
- Registering still requires an additional email confirmation loop

Email screenshot:
screenshot_2019-02-22 nathan levesque invited you to evil dinosaurs 2 - rhysyngsun gmail com - gmail

#### What GIF best describes this PR or how it makes you feel?
(Optional)

### CODE REVIEW

**Comment by rhysyngsun** [comment]
@pdpinch email screenshot above. The snackbar message for an email invitation is "Successfully invited ${email} as a contributor". Otherwise UI is unchanged.

**Comment by noisecapella** [comment]
I'm seeing some missing coverage but I'm not sure if that's just codecov missing some data. I'll check again after the `redeemed` boolean problem is fixed

**Comment by noisecapella** [comment]
I'm seeing this when I click the signup button for the new user:

    web_1      | [2019-02-26 18:28:04] ERROR 29 [django.request] log.py:228 - [b3ef7baa049a] - Internal Server Error: /api/v0/register/details/
    web_1      | Traceback (most recent call last):
    web_1      |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 34, in inner
    web_1      |     response = get_response(request)
    web_1      |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 126, in _get_response
    web_1      |     response = self.process_exception_by_middleware(e, request)
    web_1      |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 124, in _get_response
    web_1      |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
    web_1      |   File "/usr/local/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    web_1      |     return view_func(*args, **kwargs)
    web_1      |   File "/usr/local/lib/python3.6/site-packages/django/views/generic/base.py", line 68, in view
    web_1      |     return self.dispatch(request, *args, **kwargs)
    web_1      |   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 495, in dispatch
    web_1      |     response = self.handle_exception(exc)
    web_1      |   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 455, in handle_exception
    web_1      |     self.raise_uncaught_exception(exc)
    web_1      |   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 492, in dispatch
    web_1      |     response = handler(request, *args, **kwargs)
    web_1      |   File "./authentication/views.py", line 65, in post
    web_1      |     serializer.save()
    web_1      |   File "./authentication/serializers.py", line 142, in save
    web_1      |     result = super().save(**kwargs)
    web_1      |   File "/usr/local/lib/python3.6/site-packages/rest_framework/serializers.py", line 214, in save
    web_1      |     self.instance = self.create(validated_data)
    web_1      |   File "./authentication/serializers.py", line 293, in create
    web_1      |     return super()._authenticate(SocialAuthState.FLOW_REGISTER)
    web_1      |   File "./authentication/serializers.py", line 96, in _authenticate
    web_1      |     user = backend.continue_pipeline(partial)
    web_1      |   File "/usr/local/lib/python3.6/site-packages/social_core/backends/base.py", line 201, in continue_pipeline
    web_1      |     **partial.kwargs)
    web_1      |   File "/usr/local/lib/python3.6/site-packages/social_django/strategy.py", line 107, in authenticate
    web_1      |     return authenticate(*args, **kwargs)
    web_1      |   File "/usr/local/lib/python3.6/site-packages/django/contrib/auth/__init__.py", line 73, in authenticate
    web_1      |     user = backend.authenticate(request, **credentials)
    web_1      |   File "/usr/local/lib/python3.6/site-packages/social_core/backends/base.py", line 80, in authenticate
    web_1      |     return self.pipeline(pipeline, *args, **kwargs)
    web_1      |   File "/usr/local/lib/python3.6/site-packages/social_core/backends/base.py", line 83, in pipeline
    web_1      |     out = self.run_pipeline(pipeline, pipeline_index, *args, **kwargs)
    web_1      |   File "/usr/local/lib/python3.6/site-packages/social_core/backends/base.py", line 113, in run_pipeline
    web_1      |     result = func(*args, **out) or {}
    web_1      |   File "./authentication/pipeline/invite.py", line 40, in resolve_outstanding_channel_invites
    web_1      |     .prefetch_related("channel")
    web_1      |   File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 268, in __iter__
    web_1      |     self._fetch_all()
    web_1      |   File "/usr/local/lib/python3.6/site-packages/nplusone/ext/django/patch.py", line 295, in fetch_all
    web_1      |     original_fetch_all(self)
    web_1      |   File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 1186, in _fetch_all
    web_1      |     self._result_cache = list(self._iterable_class(self))
    web_1      |   File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 54, in __iter__
    web_1      |     results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
    web_1      |   File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1052, in execute_sql
    web_1      |     sql, params = self.as_sql()
    web_1      |   File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 493, in as_sql
    web_1      |     raise TransactionManagementError('select_for_update cannot be used outside of a transaction.')
    web_1      | django.db.transaction.TransactionManagementError: select_for_update cannot be used outside of a transaction.

**Comment by rhysyngsun** [comment]
@noisecapella updated with proper transactions around resolving the invites, this should ensure those don't get resolved if the reddit calls fail.

**Comment by noisecapella** [comment]
Thanks, I can confirm that functionality works now

File: authentication/pipeline/invite.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        user (User): the current user
+        is_new (bool): True if the user just got created
+    """
+    if not is_new:
+        return {}
+
+    admin_api = get_admin_api()
+
+    # resolve all channel invitations by adding the user as a contributor and a subscriber
+    for invite_id in ChannelInvitation.objects.filter(
```
Comment (noisecapella) [inline authentication/pipeline/invite.py]: Looking at coverage, it seems like this function is being run but it's not finding any items which match in the for loop and so the inner lines are never executed. Can you look into this?

File: authentication/pipeline/invite_test.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
+"""Invite pipeline tests"""
+import pytest
+
+from authentication.pipeline import invite as invite_actions
+from channels.factories.models import ChannelInvitationFactory
+
+
+@pytest.mark.parametrize("is_new", [True, False])
+@pytest.mark.parametrize("num_invites", [0, 1, 3])
+def resolve_outstanding_channel_invites(mocker, user, is_new, num_invites):
```
Comment (noisecapella) [inline authentication/pipeline/invite_test.py]: I think these need to start with `test_` to run

File: mail/templates/invite/body.html [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+<!-- 1 Column Text + Button : BEGIN -->
+  <tr>
+      <td style="background-color: #ffffff;">
+          <table role="presentation" cellspacing="0" cellpadding="0" border="0" width="100%">
+              <tr>
+                  <td style="padding: 20px; font-family: sans-serif; font-size: 15px; line-height: 20px; color: #555555;">
+                      <h1 style="margin: 0 0 10px; font-size: 25px; line-height: 30px; color: #03152d; font-weight: bold;">
+                        {{ invite.inviter.profile.name }} invited you to become a member of the {{ site_name }} channel {{ invite.channel.title }}
+                      </h1>
+                      <p style="margin: 0 0 10px;">{{ site_name }} is a place where learners, teachers and scientist worldwide meet with MIT.</p>
```
Comment (pdpinch) [inline mail/templates/invite/body.html]:
```suggestion
                      <p style="margin: 0 0 10px;">{{ site_name }} is a place where learners, teachers and researchers worldwide meet with MIT.</p>
```

**Comment by rhysyngsun** [comment]
@pdpinch it's the same style as all our other emails as it's a common header, but it renders wonky in email clients locally because all of them have reverse proxies for images for privacy reasons which can't access my localhost.

**Comment by rhysyngsun** [comment]
@noisecapella all set now

File: channels/serializers/invites.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+from channels.models import ChannelInvitation
+from channels import tasks
+
+
+class ChannelInvitationSerializer(serializers.ModelSerializer):
+    """Serializer for channel invitations"""
+
+    def create(self, validated_data):
+        with transaction.atomic():
```
Comment (noisecapella) [inline channels/serializers/invites.py]: It looks like this code is uncovered by tests. Can you add one or more tests to cover this?

File: channels/tasks.py [hunk 5]
```diff
@@ -344,3 +347,32 @@ def populate_all_posts_and_comments(self):
     )
 
     raise self.replace(results)
+
+
+@app.task
+def send_invitation_email(channel_invitation_id):
```
Comment (noisecapella) [inline channels/tasks.py]: Can you add a test for this?

File: channels/views/invites.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
+"""Channel invitations views"""
+from rest_framework import viewsets
+
+from channels.models import Channel, ChannelInvitation
+from channels.serializers.invites import ChannelInvitationSerializer
+
+
+class ChannelInvitationViewSet(viewsets.ModelViewSet):
```
Comment (noisecapella) [inline channels/views/invites.py]: Can you add one or more tests to cover the functionality in this file?

Comment (rhysyngsun) [reply channels/views/invites.py]: This viewset also needs to be protected with permissions which I missed.

**Comment by rhysyngsun** [comment]
@noisecapella all set

**Pull request merged**